ADAP-768: Update docs for materialized views/dynamic tables in v1.6 (#3935)
fishtown-analytics/docs.getdbt.com: website/docs/docs/build/materializations.md

    * are only used in one or two downstream models, and
    * do not need to be queried directly

### Materialized View

The `materialized view` materialization allows the creation and maintenance of materialized views
in the target database. This materialization makes use of the `on_configuration_change` config, which
aligns with the incremental nature of the namesake database object. This setting tells dbt to attempt to
make configuration changes directly to the object when possible, as opposed to completely recreating
the object to implement the updated configuration. Using `dbt-postgres` as an example, indexes can
be dropped and created on the materialized view without the need to recreate the materialized view itself.

> The `on_configuration_change` config has three settings:

Author: This section currently only pertains to materialized views (and dynamic tables for Snowflake), but could presumably apply to other materializations in the future. Should this be its own section? If so, where is the right spot?

Reviewer: I'm good with burning that bridge when we get there ;)

Reviewer 2: yeah. i think for 1.7 we should aim to have a standalone page for materialized views. but for now it can stay as a subsection of this page

- `apply` (default) &mdash; attempt to update the existing database object if possible, avoiding a complete rebuild
  - *Note:* if any individual configuration change requires a full refresh, a full refresh be performed in lieu of individual alter statements
- `continue` &mdash; allow runs to continue while also providing a warning that the object was left untouched
  - *Note:* this could result in downstream failures as those models may depend on these unimplemented changes
- `fail` &mdash; force the entire run to fail if a change is detected

Materialized views are implemented following this "drop through" life cycle:
1. If an object does not exist, create a materialized view
2. If an object exists, other than a materialized view, that object is dropped and replaced with a materialized view
3. If `--full-refresh` is supplied, replace the materialized view regardless of changes and the `on_configuration_change` setting

> 4. If there are no configuration changes, refresh the materialized view

Reviewer: Aren't we bypassing refresh in certain cases?

Author: Potentially. We attempt to run refresh as the default (against original spec, but missed in the 1.6 release). However, the refresh macro may do nothing, for example if the platform doesn't support it. I'll look through each adapter to make sure we're actually refreshing the materialized view, and update this line if there are cases where we aren't.

> 5. At this point there are configuration changes, proceed according to the `on_configuration_change` setting

Reviewer: ```suggestion
5. At this point if there are configuration changes, proceed according to the `on_configuration_change` setting
```

Author: I'm explicitly calling out that we have identified configuration changes, hence we proceed according to the `on_configuration_change` setting. I can reword that, but "if" would be misleading.

Materialized views are a combination of a view and a table, and serve use cases similar to incremental models.

* **Pros:**
  * Materialized views combine the query performance of a table with the data freshness of a view
  * Materialized views operate much like incremental materializations, however they are usually

> able to be refreshed on a regular cadence (depending on the database), forgoing the regular dbt batch refresh

Reviewer 3: ```suggestion
able to be refreshed without manual interference on a regular cadence (depending on the database), forgoing the regular dbt batch refresh
```

> able to be refreshed on a regular cadence (depending on the database), forgoing the regular refresh
> required with incremental materializations

Reviewer 3: ```suggestion
able to be refreshed on a regular cadence (depending on the database), forgoing the regular dbt batch refresh
required with incremental materializations
```

  * `dbt run` on materialized views correspond to a code deployment, just like views
* **Cons:**

>   * Due to the fact that materialized views are more complex database objects, they tend to have

Reviewer 3: What does this mean? What are the missing configuration options?  Can we name any specifics?

Author: The configuration options vary by platform. For instance, there's a "Limitations" section [here](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html) that talks about Redshift materialized views; of note, temporary tables and traditional views may not be used in the query. For BigQuery there's [this](https://cloud.google.com/bigquery/docs/materialized-views-create); of note, any outer joins are not allowed, nor are window functions. We link to these pages in the adapter specific docs, but I don't think we call out that there are limitations. I wouldn't want to duplicate what's on the platform's docs as that information might change, but we could at least call users' attention to the fact that these exist.

Reviewer 3: Ah I see. I read this as dbt limitations so I think we should just call out that there might be data platform limitations but go to your own docs to show. Calling this out here associated limitations to dbt

Author: @amychen1776 I updated the verbiage. Does this address your concerns? I still want to keep it in because it's one of the major factors to consider (e.g. you can't always swap incremental table code to materialized views); however, I tried to make it clear that it's a data platform limitation and not a dbt limitation.

less configuration options
  * Materialized views may not be supported by every database platform
* **Advice:**
    * Consider materialized views for use cases where incremental models are sufficient, 

> but you want to set the refresh to a schedule instead of running dbt

Reviewer 3: ```suggestion
but you would like the data platform to manage the incremental logic and refresh.
```
